Multithreaded BGPD

qlyoung opened the pull request:
This patch implements multithreading for BGP keepalive generation and network I/O.

This is not a performance oriented patch set, although it does open the door for performance improvements in the future. The intent here is to greatly improve session stability and avoid session flaps when running at large scale. Without multithreading, long-running tasks such as UI commands or large routing table updates could cause sessions to flap since they blocked all network I/O for the duration of execution. This patch solves that issue; there should be no problems running the minimum allowable hold timer of 3 seconds under "reasonably" large scale. I specifically tested this up to 800 peers, although it should theoretically scale well beyond that.

--------------------

Keepalives 
------------ 
Implementation is in `bgp_keepalives.[ch]`. This is essentially a hash table with a mutex that holds peers who need keepalives and the time until they need one. To turn on keepalives for a peer call `bgp_keepalives_on(peer)`, which wakes the thread up, inserts the peer into the table, recalculates the smallest time until any one peer needs a KA and sleeps for that long. When the time has elapsed it wakes up, generates keepalives for as many peers as it can, calculates the new sleep time in the same pass, and goes back to sleep. Generating a keepalive is just making the packet and pushing it onto `peer->obuf`, after that the I/O thread handles the rest. 

I/O
---
Implementation is in `bgp_io.[ch]`. Thread runs its own event loop and communicates with the main thread by scheduling and receiving events. When a packet is pushed onto a peer's output buffer, `bgp_writes_on(peer)` is called from the owning  which pushes an event onto the I/O pthread's event queue, causing the peer socket to be added to the list of `poll()`'d file descriptors. At some point in the future the socket will become ready for writing (or the connection will die) at which point the output buffer is serviced. As many packets as possible are written either until the write quanta is reached or until a socket write becomes blocking. In either case the peer socket is rescheduled for writing until all packets on the output buffer have been written. Upon error a `BGP_Stop` event is generated on the main thread, write events are cancelled and session teardown proceeds as per usual. If the peer exhausts all the packets on its output buffer a new write event is not scheduled, `bgp_writes_on(peer)` must be called again to reschedule it. 

Reads operate differently, when a caller wants to receive packets from a peer's socket, they register the peer with `bgp_reads_on(peer)`. When the peer's socket becomes ready for reading, as much data as possible is read off the socket into a raw input buffer. It is then broken up into individual BGP messages, whose headers are validated before being pushed onto the packet input queue. When new data is available on this queue an event is generated on the main pthread that schedules `bgp_process_packets()`, which handles processing the actual content of the messages, updating any internal state, generating FSM events etc. Reads automatically reschedule forever until `bgp_reads_off()` is called, or an error is encountered. If an error is encountered rescheduling stops and a `BGP_Stop` event is generated on the main pthread. 

Some changes to the structure of packet processing routines have been made to clean up their return codes, improve documentation and improve organization.

louberger commented:
Will need to investigate test failures.  

Valgrind loss reports:

> ==547== HEAP SUMMARY:
==547==     in use at exit: 37 bytes in 2 blocks
==547==   total heap usage: 238,429 allocs, 238,427 frees, 11,834,354 bytes allocated
==547== 
==547== Searching for pointers to 2 not-freed blocks
==547== Checked 415,944 bytes
==547== 
==547== 15 bytes in 1 blocks are definitely lost in loss record 1 of 2
==547==    at 0x4A06A2E: malloc (vg_replace_malloc.c:270)
==547==    by 0x3ACC480ED1: strdup (in /lib64/libc-2.12.so)
==547==    by 0x4C224E: qstrdup (memory.c:80)
==547==    by 0x4D460A: thread_master_create (thread.c:363)
==547==    by 0x4BB909: frr_pthread_new (frr_pthread.c:90)
==547==    by 0x4214AE: bgp_init (bgpd.c:7414)
==547==    by 0x420462: main (bgp_main.c:396)
==547== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   fun:malloc
   fun:strdup
   fun:qstrdup
   fun:thread_master_create
   fun:frr_pthread_new
   fun:bgp_init
   fun:main
}
==547== 22 bytes in 1 blocks are definitely lost in loss record 2 of 2
==547==    at 0x4A06A2E: malloc (vg_replace_malloc.c:270)
==547==    by 0x3ACC480ED1: strdup (in /lib64/libc-2.12.so)
==547==    by 0x4C224E: qstrdup (memory.c:80)
==547==    by 0x4D460A: thread_master_create (thread.c:363)
==547==    by 0x4BB909: frr_pthread_new (frr_pthread.c:90)
==547==    by 0x4214C7: bgp_init (bgpd.c:7416)
==547==    by 0x420462: main (bgp_main.c:396)
==547== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   fun:malloc
   fun:strdup
   fun:qstrdup
   fun:thread_master_create
   fun:frr_pthread_new
   fun:bgp_init
   fun:main
}
==547== LEAK SUMMARY:
==547==    definitely lost: 37 bytes in 2 blocks
==547==    indirectly lost: 0 bytes in 0 blocks
==547==      possibly lost: 0 bytes in 0 blocks
==547==    still reachable: 0 bytes in 0 blocks
==547==         suppressed: 0 bytes in 0 blocks
==547==

qlyoung commented:
@louberger #1149

louberger commented:
as mentioned in slack: when a better route is learned, the worse route is no longer being withdrawn (it is without this change)

louberger commented:
@qlyoung so what did you change in the latest push?

qlyoung commented:
@louberger https://github.com/FRRouting/frr/pull/1145/commits/2f92f6151626bf7eab5b3a6c63ac4bafbac19e5b
changes in bgp_updgrp_adv.c were missing from that commit

qlyoung commented:
rebase to resolve merge conflicts

mwinter-osr commented:
Attached are full compliance results of this pull request compared against the base sha from master branch.

There are a few new failures which need to be addressed.

[BGP4_extended_results.pdf](https://github.com/FRRouting/frr/files/1312657/BGP4_extended_results.pdf)
[BGPPLUS_extended_results.pdf](https://github.com/FRRouting/frr/files/1312658/BGPPLUS_extended_results.pdf)

mwinter-osr commented:
## Compliance Test BGP4-13.2

### Message Header Error Handling
#### RFC RFC4271, Sect. 6.1, p 31, Message Header error handling

         If the Length field of the message header is less than 19 or
         greater than 4096 then the Error Subcode MUST be set to Bad Message
         Length. The Data field MUST contain the erroneous Length field.

The response is sent, but with a wrong length field. It looks to me like a byte ordering issue:

	---- BGP4 Notification Message ----
	External peer <PEER-AS2-A> received expected BGP4 Notification Message from DUT
	Received Notification message correctly contains Error
	Subcode field set to 2 (Bad Message Length)
	ERROR: Received Notification message contains incorrect
	ERROR: value of Data field
	ERROR: Expected Value = 0012
	ERROR: Received Value = Length = 1200

mwinter-osr commented:
## Compliance Test BGP4-19.1

### Connection collision detection
#### RFC4271, Sect. 6.8, p 36, Connection collision detection

         Connection Collision Detection
         In case when a connection collision is detected, if the value of the
         local BGP Identifier is less than the remote one, the local system
         closes BGP Connection that already exists (the one that is already in
         the OpenConfirm state), and accepts BGP4 Connection initiated by the
         remote system.

This seems to be a weird response to a `neighbor x.x.x.x activate` command. The test
system uses this to trigger the DUT to start a BGP connection. In this case, FRR already
sent the TCP open and in response to this command sends a Notification (admin shut)
and starts a new BGP connection.
Not sure if this is a timing issue based on the changes. But maybe sending a `admin shut`
in response to a `neighbor x.x.x.x activate` might be a bad choice too.

mwinter-osr commented:
Added extra CI test temporary to Pull Requests to test for main failing issue.

I believe most (or potentially all) remaining issues is all based on same failure. Please ping me after this one is fixed and I'll do a full test again.

Remaining issue is FRR failing to open a BGP sessions from it's own side. To reproduce, connect a FRR node to a simple remote Box with netcat (or something else listening on port 179). Then configure BGP session to this box. The issue is FRR to open a TCP session, but never sending the BGP OPEN message.

Here is my setup:

    +---------------+             +-----------------+
    |   FRR    eth2 +-------------+ eth2  Linux box |
    +---------------+ .101     .1 +-----------------+
                   (192.168.1.0/24)

Linux box:

    ifconfig eth2 192.168.1.1/24 up
    nc -4 -k -l 179 &
    tcpdump -i eth2 port 179
	
FRR:

    conf t
    interface eth2
     ip address 192.168.1.101/24
    !
    router bgp 500
     network 192.168.1.0/24
     neighbor 192.168.1.1 remote-as 501
     neighbor 192.168.1.1 timers 60 180
     neighbor 192.168.1.1 ebgp-multihop
    end

The BGP log shows some unusual `TCP_fatal_error`:

    2017/09/21 01:09:09 BGP: bgpd 3.1-dev-20170921033945-git.73bf989 starting: vty@2605, bgp@<all>:179
    2017/09/21 01:09:56 BGP: vty[??]@> enable
    2017/09/21 01:10:05 BGP: vty[??]@# conf t
    2017/09/21 01:10:05 BGP: vty[??]@(config)# end
    2017/09/21 01:10:05 BGP: vty[??]@# configure terminal
    2017/09/21 01:10:05 BGP: vty[??]@(config)# router bgp 500
    2017/09/21 01:10:05 BGP: Creating Default VRF, AS 500
    2017/09/21 01:10:05 BGP: Registering VRF 0
    2017/09/21 01:10:05 BGP: Rx Router Id update VRF 0 Id 192.168.123.167/32
    2017/09/21 01:10:05 BGP: Rx Intf add VRF 0 IF br1
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF br1 addr 1.1.1.2/24
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF br1 addr fe80::21e:99ff:fe99:9900/64
    2017/09/21 01:10:05 BGP: Rx Intf add VRF 0 IF eth0
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth0 addr 192.168.123.167/23
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth0 addr 2001:470:113:300:0:aaaa:d:31/64
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth0 addr fe80::21c:42ff:fe22:2231/64
    2017/09/21 01:10:05 BGP: Rx Intf add VRF 0 IF eth1
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth1 addr fe80::21e:99ff:fe99:9900/64
    2017/09/21 01:10:05 BGP: Rx Intf add VRF 0 IF eth2
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth2 addr 192.168.1.101/24
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth2 addr fe80::21e:99ff:fe99:9901/64
    2017/09/21 01:10:05 BGP: Rx Intf add VRF 0 IF eth3
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth3 addr 192.168.2.101/24
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth3 addr fe80::21e:99ff:fe99:9902/64
    2017/09/21 01:10:05 BGP: Rx Intf add VRF 0 IF eth4
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth4 addr 192.168.3.101/24
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth4 addr fe80::21e:99ff:fe99:9903/64
    2017/09/21 01:10:05 BGP: Rx Intf add VRF 0 IF eth5
    2017/09/21 01:10:05 BGP: Rx Intf address add VRF 0 IF eth5 addr fe80::21e:99ff:fe99:9904/64
    2017/09/21 01:10:05 BGP: Rx Intf add VRF 0 IF lo
    2017/09/21 01:10:05 BGP: Rx Intf add VRF 0 IF tap0
    2017/09/21 01:10:05 BGP: vty[??]@(config-router)# network 192.168.1.0/24
    2017/09/21 01:10:05 BGP: vty[??]@(config-router)# neighbor 192.168.1.1 remote-as 501
    2017/09/21 01:10:05 BGP: vty[??]@(config-router)# neighbor 192.168.1.1 timers 60 180
    2017/09/21 01:10:05 BGP: vty[??]@(config-router)# neighbor 192.168.1.1 ebgp-multihop
    2017/09/21 01:10:05 BGP: vty[??]@(config-router)# end
    2017/09/21 01:10:06 BGP: 192.168.1.1 [FSM] Timer (start timer expire).
    2017/09/21 01:10:06 BGP: 192.168.1.1 [FSM] BGP_Start (Idle->Connect), fd -1
    2017/09/21 01:10:06 BGP: Allocated bnc 192.168.1.1/32 peer 0x17ee080
    2017/09/21 01:10:06 BGP: 192.168.1.1 [Event] Connect start to 192.168.1.1 fd 22
    2017/09/21 01:10:06 BGP: 192.168.1.1 [FSM] Non blocking connect waiting result, fd 22
    2017/09/21 01:10:06 BGP: 192.168.1.1 went from Idle to Connect
    2017/09/21 01:10:06 BGP: 0: Rcvd NH update 192.168.1.1/32 - metric 0/0 #nhops 0/0 flags 0x2
    2017/09/21 01:10:06 BGP: NH update for 192.168.1.1/32 - flags 0x2 chgflags 0x0 - evaluate paths
    2017/09/21 01:10:06 BGP: evaluate_paths: Updating peer (192.168.1.1) status with NHT
    2017/09/21 01:10:06 BGP: 192.168.1.1 [FSM] TCP_fatal_error (Connect->Idle), fd 22
    2017/09/21 01:10:06 BGP: 192.168.1.1 went from Connect to Idle
    2017/09/21 01:10:06 BGP: 0: Rcvd NH update 192.168.1.1/32 - metric 0/0 #nhops 1/0 flags 0xa
    2017/09/21 01:10:06 BGP:     nhop via if 4
    2017/09/21 01:10:06 BGP: NH update for 192.168.1.1/32 - flags 0x3 chgflags 0x1 - evaluate paths
    2017/09/21 01:10:06 BGP: evaluate_paths: Updating peer (192.168.1.1) status with NHT
    2017/09/21 01:10:06 BGP: 192.168.1.1 [FSM] BGP_Start (Idle->Connect), fd -1
    2017/09/21 01:10:06 BGP: 192.168.1.1 [Event] Connect start to 192.168.1.1 fd 22
    2017/09/21 01:10:06 BGP: 192.168.1.1 [FSM] Non blocking connect waiting result, fd 22
    2017/09/21 01:10:06 BGP: 192.168.1.1 went from Idle to Connect
    2017/09/21 01:10:07 BGP: vty[??]@# exit

qlyoung commented:
@mwinter-osr please re-run your compliance tests

mwinter-osr commented:
Much better this time, but still a few issues. Attached are the test reports. Please be aware that I've compared to base sha 5cf0c43 (which is the one you originally submitted against), but the latest update of the PR is now on top of sha 9650716. There is a slight chance that the failures are based on changes in master between these commits. (I don't have a run from sha 9650716 at this time, so would need some time to do this if needed)

[BGP-AS4_extended_results.pdf](https://github.com/FRRouting/frr/files/1332091/BGP-AS4_extended_results.pdf)
[BGP4_extended_results.pdf](https://github.com/FRRouting/frr/files/1332093/BGP4_extended_results.pdf)
[BGPPLUS_extended_results.pdf](https://github.com/FRRouting/frr/files/1332092/BGPPLUS_extended_results.pdf)
[BGPPLUS-AS4_extended_results.pdf](https://github.com/FRRouting/frr/files/1332090/BGPPLUS-AS4_extended_results.pdf)

mwinter-osr commented:
## Outstanding issue: ANVL-BGP4-19.6

### Connection Collision Detection
Closing the BGP4 Connection (that results from the collision resolution
procedure) is accomplished by sending the NOTIFICATION message with
the Error Code Cease.

### Reference:
RFC4271, Sect. 6.8, p 36, Connection collision detection

This seems to be unreliable. Out of 4 attempts, it worked correctly 2 times and failed 2 times. When it failed, it looked like FRR opened 2 TCP connections and none of them got closed.

Sending log, pcap etc details to @qlyoung

mwinter-osr commented:
## Outstanding issue: ANVL-BGPPLUS-11.1

(Testcase is probably not relevant here - the failure seems to be related to some timing/corruption)

### Message Formats
The maximum message size is 4096 octets. All implementations are
required to support this maximum message size.

### Reference
RFC 4271, Sect. 4, p 10, Message Formats

The issue in this failure (1 failure out of 4 attempts) looks like some timing or corruption. The BGP session is opened, but just after the open, FRR closes it with a notification `Message Header Error/Connection Not Synchronized` message.

Sending log, pcap etc details to @qlyoung

mwinter-osr commented:
## Outstanding issue: ANVL-BGPPLUS-27.6

This is the same as ANVL-BGP4-19.6, but for IPv6 instead of IPv4. Assumed to be same issue
(ask if you want the details anyway)

mwinter-osr commented:
## Outstanding Issue: ANVL-BGPPLUS-29.7

### BGP Finite State Machine
In state OpenSent if the Hold Timer expires, the local system sends
NOTIFICATION message with Error Code Hold Timer Expired.

### Reference
NEGATIVE RFC 4271, Sect. 8.2.2, p 63, BGP Finite State machine

ANVL opens a BGP session to Open Sent mode and sits idle, waiting for HOLD time expiration notification. (In my testcase, the timeout is configured to 90s). ANVL waits for 100s without sending a Keepalive, but does not receive a notification.
Issue failed 1 out of 4 attempts.

Sending log, pcap etc details to @qlyoung

qlyoung commented:
Regarding 19.6 and 27.6, my assessment is that this is the same problem described in #1219 and I will not be addressing that in this pull request.
Regarding 29.7 I have not been able to reproduce this after a few hundred attempts.
Regarding 11.1 I have not yet been able to reproduce this either.

Can you give me details on your environment?

qlyoung commented:
29.7 is a problem in ANVL. The hold timer is set to the global default of 180s. ANVL expects that we use the hold timer specified for the peer. The RFC merely says to use "a large value" and suggests 4 minutes. Therefore ANVL is testing unspecified behavior. I will not be fixing this test and suggest we modify or ignore it.

qlyoung commented:
Rebased to fix merge conflicts.

I have rerun ANVL BGPPLUS 11.1 (4096-byte message causing unsynchronized header notification) ~15 times and have yet to see this reproduce under the same test environment reported here.

qlyoung commented:
@mwinter-osr please take a look at the testing related comments above and update your review

mwinter-osr commented:
I'm going to rerun the complete set of BGP tests with ANVL. Will report back with the results in approx 24hrs

mwinter-osr commented:
Crashed with Address Sanitizer during ANVL-BGP4-18.4:

From the bgpd.log:

	BGP: [Event] BGP connection from host 192.168.1.1 fd 19
	BGP: 192.168.1.1 went from Idle to Active
	BGP: 192.168.1.1 [FSM] TCP_connection_open (Active->OpenSent), fd 19
	BGP: 192.168.1.1 passive open
	BGP: 192.168.1.1 Sending hostname cap with hn = comp1-dut, dn = (null)
	BGP: 192.168.1.1 sending OPEN, version 4, my as 500, holdtime 180, id 192.168.1.101
	BGP: 192.168.1.1 went from Active to OpenSent
	BGP: 192.168.1.1 rcv OPEN, version 4, remote-as (in open) 501, holdtime 90, id 192.168.1.1
	BGP: 192.168.1.1 rcvd OPEN w/ OPTION parameter len: 0
	BGP: 192.168.1.1 [FSM] Receive_OPEN_message (OpenSent->OpenConfirm), fd 19
	BGP: 192.168.1.1 sending KEEPALIVE
	BGP: 192.168.1.1 went from OpenSent to OpenConfirm
	BGP: 192.168.1.1 KEEPALIVE rcvd
	BGP: 192.168.1.1 [FSM] Receive_KEEPALIVE_message (OpenConfirm->Established), fd 19
	BGP: 192.168.1.1: peer transfer 0x629000f32200 fd 19 -> 0x629000f19200 fd -1)
	BGP: 192.168.1.1 went from OpenConfirm to Established
	BGP: [Event] Deleting stub connection for peer 192.168.1.1
	BGP: 192.168.1.1 went from Active to Deleted
	BGP: 192.168.1.1 [FSM] Timer (routeadv timer expire)
	BGP: 192.168.1.1: BGP type 3 length 32 is too large, attribute total length is 18.  attr_endp is 0x604000446f95.  endp is 0x604000446f79

From the test: The Cease NOTIFICATION message MUST NOT be used when a fatal error
indicated by this section does exist. (This test checks the case when the error is in UPDATE Message fields)
Reference: `NEGATIVE RFC4271, Sect. 6.7, p 35, Cease`

Address Sanitizer breaks with:

### ERROR: AddressSanitizer: heap-buffer-overflow on address 0x604000446f7d at pc 0x7f445c940935 bp 0x7ffff50dcd90 sp 0x7ffff50dc538

READ of size 35 at 0x604000446f7d thread T0

    #0 0x7f445c940934 in __asan_memcpy (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x8c934)
    #1 0x5eae23 in memcpy /usr/include/x86_64-linux-gnu/bits/string3.h:53
    #2 0x5eae23 in stream_write lib/stream.c:985
    #3 0x49fc90 in bgp_notify_send_with_data /home/anvl/PR1145/bgpd/bgp_packet.c:593
    #4 0x4534df in bgp_attr_parse /home/anvl/PR1145/bgpd/bgp_attr.c:2275
    #5 0x4a4009 in bgp_update_receive /home/anvl/PR1145/bgpd/bgp_packet.c:1408
    #6 0x4a4009 in bgp_process_packet /home/anvl/PR1145/bgpd/bgp_packet.c:2136
    #7 0x5f6d41 in thread_call lib/thread.c:1485
    #8 0x5ba69d in frr_run lib/libfrr.c:879
    #9 0x41f062 in main /home/anvl/PR1145/bgpd/bgp_main.c:404
    #10 0x7f445b79782f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #11 0x41e5e8 in _start (/usr/local/PR1145/sbin/bgpd+0x41e5e8)

0x604000446f7d is located 0 bytes to the right of 45-byte region [0x604000446f50,0x604000446f7d)

allocated by thread T1 here:

    #0 0x7f445c94c602 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x98602)
    #1 0x5c08e1 in qmalloc lib/memory.c:63
    #2 0x5e327e in stream_new lib/stream.c:100
    #3 0x584945 in bgp_process_reads /home/anvl/PR1145/bgpd/bgp_io.c:307
    #4 0x5f6d41 in thread_call lib/thread.c:1485
    #5 0x585016 in bgp_io_start /home/anvl/PR1145/bgpd/bgp_io.c:88
    #6 0x7f445bb486b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)

Thread T1 created by T0 here:

    #0 0x7f445c8ea253 in pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x36253)
    #1 0x5b0a1e in frr_pthread_run lib/frr_pthread.c:133
    #2 0x445e5b in bgp_pthreads_run /home/anvl/PR1145/bgpd/bgpd.c:7429
    #3 0x41f019 in main /home/anvl/PR1145/bgpd/bgp_main.c:403
    #4 0x7f445b79782f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow ??:0 __asan_memcpy

	Shadow bytes around the buggy address:
	  0x0c0880080d90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
	  0x0c0880080da0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
	  0x0c0880080db0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
	  0x0c0880080dc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
	  0x0c0880080dd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
	=>0x0c0880080de0: fa fa fa fa fa fa fa fa fa fa 00 00 00 00 00[05]
	  0x0c0880080df0: fa fa 00 00 00 00 00 fa fa fa fd fd fd fd fd fa
	  0x0c0880080e00: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
	  0x0c0880080e10: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
	  0x0c0880080e20: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fd
	  0x0c0880080e30: fa fa fd fd fd fd fd fa fa fa fd fd fd fd fd fa
	Shadow byte legend (one shadow byte represents 8 application bytes):
	  Addressable:           00
	  Partially addressable: 01 02 03 04 05 06 07 
	  Heap left redzone:       fa
	  Heap right redzone:      fb
	  Freed heap region:       fd
	  Stack left redzone:      f1
	  Stack mid redzone:       f2
	  Stack right redzone:     f3
	  Stack partial redzone:   f4
	  Stack after return:      f5
	  Stack use after scope:   f8
	  Global redzone:          f9
	  Global init order:       f6
	  Poisoned by user:        f7
	  Container overflow:      fc
	  Array cookie:            ac
	  Intra object redzone:    bb
	  ASan internal:           fe

mwinter-osr commented:
## ANVL-BGP4-5.1 Fails 50% of the tests (2 fails out of 4 testruns)

### Reference:
RFC4271, Sect. 4.4, p 21, KEEPALIVE Message Format

### Summary: 
KEEPALIVE messages MUST NOT be sent more frequently than one per
second. The Hold Time MUST be either zero or at least three seconds

	ERROR: Time interval between the reception of two successive
	ERROR: BGP4 KeepAlives is incorrect
	ERROR: Expected Value = Greater than or Equal to 1.0 seconds
	ERROR: Received Value = 0.998678 second

I think this one can be ignored. This is close enough to 1 sec.

mwinter-osr commented:
## ANVL-BGP4-31.6 Fails 50% of the time (2 fails out of 4 testruns)

### Reference: 
RFC4271, Sect. 9.2.2.2, p 87, Aggregating Routing Information

### Summary:
Aggregating Routing Information:
If at least one route among routes that are aggregated has ORIGIN with
the value EGP, then the aggregated route must have the ORIGIN
attribute with the value EGP.

ANVL announces the same route twice in short order 30.0.5.128/25 with Origin IGP and immediately afterwards (0.1sec later) 30.0.5.192/26 with Origin EGP. The DUT is configured to aggregate to /24 and ANVL expects the resulting Origin to be EGP.
FRR sends it with IGP first (very fast response to 1st update) and then followed by the update for EGP.
I think this part is acceptable.

However I'm concerned as the 2nd BGP packet contains 2 update messages - one for announce and one for withdraw of the same route. Not sure if this is defined in which order they get processed (correct for this case would be to ignore the withdraw)

	UPDATE Message, UPDATE Message
	Frame 27: 139 bytes on wire (1112 bits), 139 bytes captured (1112 bits)
	Ethernet II, Src: Vantanol_99:99:02 (00:1e:99:99:99:02), Dst: EleluxIn_00:00:02 (00:1e:
	11:00:00:02)
	Internet Protocol Version 4, Src: 192.168.2.101, Dst: 192.168.2.1
	Transmission Control Protocol, Src Port: 179 (179), Dst Port: 16590 (16590), Seq: 217, Ack: 49,
	Len: 85
	Border Gateway Protocol - UPDATE Message
		Marker: ffffffffffffffffffffffffffffffff
		Length: 58
		Type: UPDATE Message (2)
		Withdrawn Routes Length: 0
		Total Path Attribute Length: 31
		Path attributes
			Path Attribute - ORIGIN: EGP
			Path Attribute - AS_PATH: 500
			Path Attribute - NEXT_HOP: 192.168.2.101
			Path Attribute - ATOMIC_AGGREGATE
			Path Attribute - AGGREGATOR: AS: 500 origin: 192.168.1.101
		Network Layer Reachability Information (NLRI)
			30.0.5.0/24
				NLRI prefix length: 24
				NLRI prefix: 30.0.5.0
	Border Gateway Protocol - UPDATE Message
		Marker: ffffffffffffffffffffffffffffffff
		Length: 27
		Type: UPDATE Message (2)
		Withdrawn Routes Length: 4
		Withdrawn Routes
			30.0.5.0/24
				Withdrawn route prefix length: 24
				Withdrawn prefix: 30.0.5.0
		Total Path Attribute Length: 0

qlyoung commented:
Please run ANVL on master and report if this fails there as well.

mwinter-osr commented:
Already did (see ANVL results attached earlier) and it doesn't happen there.

However as this only fails sometimes, it might be just luck and the timing change of the multithreading exposing the problem

mwinter-osr commented:
False Alarm!

The issue reported by ANVL-BGP4-31.6 is NOT in the code, but in the way ANVL tests.

Here is what happened:

    +-------------+        +---------+        +-------------+
    | ANVL peer 1 +--------+ FRR DUT +--------+ ANVL peer 2 |
    +-------------+        +---------+        +-------------+

FRR DUT is configured to aggregate 30.0.5.0/24 (`aggregate-address 30.0.5.0/24 summary-only`).

1) ANVL peer 1 sends 30.0.5.128/25 with origin IGP
2) FRR DUT sends the aggregation 30.0.5.0/24 with origin IGP to ANVL peer 2
3) ANVL peer 1 sends 30.0.5.192/26 with origin EGP
4) ANVL peer 2 now expects 30.0.5.0/24 with origin EGP to show up.

However, FRR seems to send the update < 0.1 sec after step 1 (before the 2nd update from
ANVL shows up) and at this time correctly with origin IGP
Now in step 4, ANVL reads on peer2, but picks up the update from before (which it didn't process yet) and thinks it got it wrong. ANVL now kills peer1 (with notification sent).

The weird update I reported on is getting triggered at this time: The update to origin EGP and the withdraw from ANVL shutting the peer 1 down.

This is a ANVL bug. Ignore for now and I'll get ANVL fixed.

mwinter-osr commented:
From the functionality the PR is all good. I have NOT tested any performance.

qlyoung commented:
This latest update is a rebase + 7 commits to address performance issues. They should not have a functional impact on RFC compliance / ANVL test results.

I have also updated the PR description (first comment) to explain the intent of this PR, since I failed to do so initially.

donaldsharp commented:
@louberger @mwinter-osr anything holding this up?

donaldsharp commented:
Once this is rebased, another compliance test run?

qlyoung commented:
Rebased, fixed `make check` failure introduced with last rebase, added some tweaks based on performance testing results.

For interested parties, here is a brief performance breakdown focused on convergence time analysis.

-------------
At small scale, convergence times have increased slightly due to some additional overhead and intrinsic timing changes:
![frr](https://user-images.githubusercontent.com/6827003/33458171-29638f9a-d5f3-11e7-9e6b-81a9b661696f.png)

However, as scale increases, convergence time reduces dramatically:
![frr](https://user-images.githubusercontent.com/6827003/33458219-5d14e514-d5f3-11e7-93a1-e28383047501.png)

As demonstrated in the graph above, per-peer memory usage has also been reduced; the second graph exhibits a fivefold decrease in memory consumption during convergence.

-------------

@FRRouting/maintainers In my opinion this PR is ready to merge.
